hw/drivers/crypto/crypto_da1469x/src/crypto_da1469x.c

```diff
@@ -0,0 +1,236 @@
+        mask = 2 << AES_HASH_CRYPTO_CTRL_REG_CRYPTO_AES_KEY_SZ_Pos;
+        break;
+    }
+    AES_HASH->CRYPTO_CTRL_REG |= mask;
+
+    /* activate key expansion */
+    AES_HASH->CRYPTO_CTRL_REG |= AES_HASH_CRYPTO_CTRL_REG_CRYPTO_AES_KEXP_Msk;
+
+    AES_HASH->CRYPTO_KEYS_START = (uint32_t)key;
+    AES_HASH->CRYPTO_FETCH_ADDR_REG = (uint32_t)inbuf;
+    AES_HASH->CRYPTO_DEST_ADDR_REG = (uint32_t)outbuf;
+    /* NOTE: length register allows only 24-bits */
+    AES_HASH->CRYPTO_LEN_REG = len;
+
+    if (op == CRYPTO_OP_ENCRYPT) {
+        AES_HASH->CRYPTO_CTRL_REG |= AES_HASH_CRYPTO_CTRL_REG_CRYPTO_ENCDEC_Msk;
+    } else {
+        AES_HASH->CRYPTO_CTRL_REG &= ~AES_HASH_CRYPTO_CTRL_REG_CRYPTO_ENCDEC_Msk;
+    }
+
+    /* start encryption/decryption */
+    AES_HASH->CRYPTO_START_REG |= AES_HASH_CRYPTO_START_REG_CRYPTO_START_Msk;
+
+    /*
+     * Wait to finish.
+     *
+     * FIXME: refactor busy wait!?
+     */
+    while (!(AES_HASH->CRYPTO_STATUS_REG &
+                AES_HASH_CRYPTO_STATUS_REG_CRYPTO_INACTIVE_Msk));
+
+    /*
+     * Update crypto framework internals
+     */
+    switch (mode) {
+    case CRYPTO_MODE_CBC:
+        if (op == CRYPTO_OP_ENCRYPT) {
+            memcpy(iv, &outbuf[len-AES_BLOCK_LEN], AES_BLOCK_LEN);
+        } else {
+            memcpy(iv, iv_save, AES_BLOCK_LEN);
+        }
+        break;
+    case CRYPTO_MODE_CTR:
+        inc = (len + AES_BLOCK_LEN - 1) / AES_BLOCK_LEN;
+        carry = 0;
+        for (i = AES_BLOCK_LEN; (inc != 0 || carry != 0) && i > 0; --i) {
+            v = carry + (uint8_t)inc + iv[i - 1];
+            iv[i - 1] = (uint8_t)v;
+            inc >>= 8;
+            carry = v >> 8;
+        }
+        break;
+    }
+
+    da1469x_clock_amba_disable(CRG_TOP_CLK_AMBA_REG_AES_CLK_ENABLE_Msk);
+    os_mutex_release(&gmtx);
+    return len;
+}
+
+static uint32_t
+da1469x_crypto_encrypt(struct crypto_dev *crypto, uint16_t algo, uint16_t mode,
+        const uint8_t *key, uint16_t keylen, uint8_t *iv, const uint8_t *inbuf,
+        uint8_t *outbuf, uint32_t len)
+{
+    return da1469x_crypto_op(crypto, CRYPTO_OP_ENCRYPT, algo, mode, key,
+            keylen, iv, inbuf, outbuf, len);
+}
+
+static uint32_t
+da1469x_crypto_decrypt(struct crypto_dev *crypto, uint16_t algo, uint16_t mode,
+        const uint8_t *key, uint16_t keylen, uint8_t *iv, const uint8_t *inbuf,
+        uint8_t *outbuf, uint32_t len)
+{
+    return da1469x_crypto_op(crypto, CRYPTO_OP_DECRYPT, algo, mode, key,
+            keylen, iv, inbuf, outbuf, len);
+}
+
+static int
+da1469x_crypto_dev_open(struct os_dev *dev, uint32_t wait, void *arg)
+{
+    struct crypto_dev *crypto;
+
+    crypto = (struct crypto_dev *)dev;
+    assert(crypto);
+
+    return 0;
+}
+
+int
+da1469x_crypto_dev_init(struct os_dev *dev, void *arg)
+{
+    struct crypto_dev *crypto;
+
+    crypto = (struct crypto_dev *)dev;
+    assert(crypto);
+
+    OS_DEV_SETHANDLERS(dev, da1469x_crypto_dev_open, NULL);
+
+    assert(os_mutex_init(&gmtx) == 0);
```

Comment:
I'm not sure if initializing mutex in assert is such a good idea

Reply:
Ok, I separated both.